### PR TITLE
 mergesort in O(1) space complexity in C

### PR DESCRIPTION
**Pull request by ag278**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

This is C code for merging 2 sorted arrays in O(1) extra space and in Nlog(N) + Mlog(M) time complexity as a part of CH20

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

Issue #909

### CODE REVIEW

**Comment by ag278** [comment]
@iamrajiv  @atarax665  @SKAUL05 @Kajol-Kumari @Abhijit2505  please have a review at this PR

**Comment by ag278** [comment]
@iamrajiv @Kajol-Kumari @siddharth25pandey @plazzy99 please review this

File: C/sort/mergesort_in_O(1)_space_complexity.c [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    
+    //taking input from the user
+    
+	int n,m;    
+	printf("please enter the number of elements for the first array\n");
+	scanf("%d ",&n);
+	printf("please enter the number of elements for the second array\n");
+	scanf("%d ",&m);
+	
+	int a[100000] , b[100000];
```
Comment (vatsalkesarwani12) [inline C/sort/mergesort_in_O(1)_space_complexity.c]: make this dynamic

File: C/sort/mergesort_in_O(1)_space_complexity.c [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            {
+                a[n - i - 1] = a[n - i - 1] - b[i];          //swapping the elements
+                b[i]= a[n - i - 1] + b[i]; 
+                a[n- i - 1] = b[i] - a[n - i - 1];
+            } 
+    }
+    
+    // using the inplace algorithm we will now sorting the individual arrays 
+    
+    //sorting first array
```
Comment (vatsalkesarwani12) [inline C/sort/mergesort_in_O(1)_space_complexity.c]: I don't this is a merge sort algorithm

**Review by vatsalkesarwani12 — CHANGES_REQUESTED** [review]

**Comment by ag278** [comment]
@plazzy99  i have made the required changes. please verify it

**Review by iamrajiv — CHANGES_REQUESTED** [review]
@ag278 Remove unnecessary line breaks and indentations.

**Comment by ag278** [comment]
@iamrajiv @plazzy99 I have made the required changes. Please verify it

**Review by iamrajiv — CHANGES_REQUESTED** [review]
@ag278 I have formatted the code according to https://google.github.io/styleguide/cppguide.html. Update like this
```
// algorithm for merging two sorted array in O(1) space complexity and in
// Nlog(N) + Mlog(M) time complexity
#include <stdio.h>
#include <stdlib.h>
int i, j;

void mergesort(int a[], int b[], int n, int m) {
  int min;
  if (n >= m) {
    min = m;
  } else {
    min = n;
  }
  for (i = 0; i < min; i++)  // we are swapping the elements of both the array
    // which should be there in the result
  {
    if (a[n - i - 1] > b[i]) {
      a[n - i - 1] = a[n - i - 1] - b[i];  // swapping the elements
      b[i] = a[n - i - 1] + b[i];
      a[n - i - 1] = b[i] - a[n - i - 1];
    }
  }
  // sorting the first array
  int x = n;
  int round1 = x / 2 + x % 2;
  while (round1 > 0 && x > 1) {
    for (int i = 1; i + round1 <= n;
         i++)  // Comparing elements in the first array
      if (a[i - 1] > a[i + round1 - 1]) {
        a[i - 1] = a[i - 1] - a[i + round1 - 1];
        a[i + round1 - 1] = a[i - 1] + a[i + round1 - 1];
        a[i - 1] = a[i + round1 - 1] - a[i - 1];
      }

    x = round1;
    round1 = x / 2 + x % 2;
  }
  // Sorting  the second array
  int y = m;
  int round2 = y / 2 + y % 2;
  while (round2 > 0 && y > 1) {
    for (i = 1; i + round2 <= m; i++)  // Comparing elements in the second array
      if (b[i - 1] > b[i + round2 - 1]) {
        b[i - 1] = b[i - 1] - b[i + round2 - 1];
        b[i + round2 - 1] = b[i - 1] + b[i + round2 - 1];
        b[i - 1] = b[i + round2 - 1] - b[i - 1];
      }

    y = round2;
    round2 = y / 2 + y % 2;
  }
}

void print(int a[], int b[], int n, int m) {
  for (i = 1; i <= n;
       i++)  // printing the elements of the first and second array
    printf("%d ", a[i - 1]);
  for (i = 1; i <= m; i++) printf("%d ", b[i - 1]);
}

int main() {
  int n, m;
  printf(
    "please enter the number of elements for the first array\n");  // taking
  // input
  // from the
  // user
  scanf("%d ", &n);
  printf("please enter the number of elements for the second array\n");
  scanf("%d ", &m);
  int capacity = 100;
  int used = 0;
  int *a = (int *)malloc(sizeof(int) * capacity);
  int *b = (int *)malloc(sizeof(int) * capacity);

  printf("please enter the elements of the first array\n");
  for (i = 1; i <= n; i++) {
    scanf("%d", &a[i - 1]);
    if (used == capacity) {
      capacity *= 2;
      a = (int *)realloc(a, sizeof(int) * capacity);
    }
    a[used] = a[i - 1];
    used++;
  }

  printf("please enter the elements of the second array\n");
  for (i = 1; i <= m; i++) {
    scanf("%d", &b[i - 1]);
    if (used == capacity) {
      capacity *= 2;
      b = (int *)realloc(b, sizeof(int) * capacity);
    }
    b[used] = b[i - 1];
    used++;
  }
  mergesort(a, b, n, m);
  print(a, b, n, m);
  return 0;
}

/*
Sample input/output
Enter size of first array:
6
Enter size of second array:
5
Enter array elements:
1 5 7 9 13 17
Enter array elements:
4 6 10 13 15
output
1 4 5 6 7 9 10 13 13 15 17
*/

```

**Comment by ag278** [comment]
@iamrajiv i have made the changes . Please verify it

File: C/sort/mergesort_in_O(1)_space_complexity.c [hunk 3]
```diff
@@ -0,0 +1,114 @@
+// algorithm for merging two sorted array in O(1) space complexity and in
+// Nlog(N) + Mlog(M) time complexity
+#include <stdio.h>
+#include <stdlib.h>
+int i, j;
+
+void mergesort(int a[], int b[], int n, int m) {
+  int min;
```
Comment (iamrajiv) [inline C/sort/mergesort_in_O(1)_space_complexity.c]: @ag278 You have imported `#include <stdlib.h>` so you can use `int minimum = min(n, m)`.

File: C/sort/mergesort_in_O(1)_space_complexity.c [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    min = m;
+  } else {
+    min = n;
+  }
+  for (i = 0; i < min; i++)  // we are swapping the elements of both the array
+    // which should be there in the result
+  {
+    if (a[n - i - 1] > b[i]) {
+      a[n - i - 1] = a[n - i - 1] - b[i];  // swapping the elements
+      b[i] = a[n - i - 1] + b[i];
```
Comment (iamrajiv) [inline C/sort/mergesort_in_O(1)_space_complexity.c]: @ag278 I think you are making both the array bitonic i.e. making the sequence of numbers which is first strictly increasing then after a point strictly decreasing so you can mention this in comments.

**Review by iamrajiv — CHANGES_REQUESTED** [review]
@ag278 Many times swapping is used in the program so you can make a separated function of swapping so that the code length gets decrease and gets easy to understand. Something like this
```c
void swap(int &x, int &y) 
{ 
    int temp = x; 
    x = y; 
    y = temp; 
} 
```

**Comment by ag278** [comment]
@iamrajiv i have made the changes. Please verify it

File: C/sort/mergesort_in_O(1)_space_complexity.c (hunk 4 above)
Comment (ag278) [reply C/sort/mergesort_in_O(1)_space_complexity.c]: @iamrajiv I am not making the arry bitonic. I am just transferring the elements fron the arrays in such a way that after swapping the elements of array both the arrays should contain only those elements which would be present in the final arrays after sorting the arrays. I am using similar type of  inplace algorithm for that

**Review by iamrajiv — APPROVED** [review]

**Comment by ag278** [comment]
@plazzy99 @Kajol-Kumari could you please review my PR. I have made the changes as you have asked

**Review by SKAUL05 — APPROVED** [review]

**Review by vatsalkesarwani12 — APPROVED** [review]

**Pull request merged**